### PR TITLE
Use local abstractions sources instead of packages

### LINKED ISSUE #86: [PROPOSAL] Add a developer guide for integration tests and usage of abstractions library
[Abstractions library](https://github.com/opensearch-project/opensearch-net-abstractions) used in integration tests and it is responsible for downloading, unpacking and installation of OpenSearch cluster.
Updating integration test infrastructure usually requires updates in abstractions, the document to be added should describe how a developer can test and debug abstractions. Default behaviour getting abstractions will be to pull down Abstractions from Nuget however a developer method is still desired.

There are few options to test abstractions locally or in GitHub Actions (GHA) environment. The document should describe at least one of them.
### 1. Add abstraction binaries to client’s repo as it happens now (as of client 1.0 release).

1. Compile abstractions
2. Build packages
3. Remove old packages
4. Deploy to the directory in .net client repo
5. Update references

#### Pros
* Easy to deploy
* Can be used to run tests locally and in GHA both

#### Cons
* Not automated process - human error may occur
* Storing compiled binaries is a bad practice
* No strict link between stored packages and source code on the repository or a local branch
* Due to limitations of abstractions versioning it is possible to get different package with the same version

### 2. Copy abstractions code into clients repo for R&D purposes

1. Replace package references by project references
2. (Optional) Commit entire abstractions lib (sources) into client's repo to have tests run in GHA
3. Apply changes done in abstractions to the abstractions repo

#### Pros
* No local deployment required
* Easy to debug

#### Cons
* Copy entire repository into another one is a bad practice even for debugging
* It is possible to have a mistake on stage of copying back and forth

### 3. Checkout abstractions and build in the workflow during GHA

There are 2 ways of running test with this method.
Using package reference (like method 1 above):
1. Checkout abstractions on the given branch or commit hash
4. Compile it
5. Deploy it to client's repo
6. Run tests

Using project reference (like method 2 above):
1. Checkout abstractions on the given branch or commit hash
2. Run tests

#### Pros
* Fully automated process
* Strict link between test results and source code

#### Cons
* This might require a GH token or something else like this - OpenSearch denied this
* Debugging is impossible
* Require R&D of GHA workflow

### 4. Have an easy way to do unofficial releases, not visible/available for users

1. Make a debug of abstractions
2. Publish it to CI nuget repository
3. Use it in .net client

#### Pros
* Fully automated process
* Strict link between test results and source code
* Tests like a real nuget package
* Could be used for debug and for GHA

#### Cons
* Infrastructure team involved
* No rapid integration
* For security reasons only maintainers will be able to use this method

### PR DESCRIPTION
**Pull request by Yury-Fridlyand**
Signed-off-by: Yury-Fridlyand <yuryf@bitquilltech.com>

### Description
[Abstractions lib sources](https://github.com/opensearch-project/opensearch-net-abstractions) were copied into .net client to use in tests. No need any more to store local nuget packages of abstractions, release them or worry about outdated/broken references to them.
Few fixes were made to let workflows pass. I have also update license headers to comply with #69/#78.

See https://github.com/Bit-Quill/opensearch-net/pull/40 for more details.

This PR follows option 2 in #86.

### Issues Resolved
#86

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: abstractions/src/Directory.Build.props
Comment (Xtansia) [inline abstractions/src/Directory.Build.props:12]: The relative path is somewhat confusing, I'm assuming due to the nature of MSBuild, this actually gets evaluated at the actual project's context, meaning this does end up resolving to `abstractions/build/keys/keypair.snk`. 

Is there much reason this couldn't just use the main SNK as used by the clients, especially given this hasn't been truly published before as far as I'm aware?

**Review by Xtansia — CHANGES_REQUESTED** [review]
From the perspective of getting these repos merged with the fewest changes, this is really good. However I'd really like to see the solutions properly reconciled into one, rather than this nesting. Whether that should be part of this or a follow up pass instead is another question.

**Review by MitchellGale-BitQuill — DISMISSED** [review]

File: MAINTAINERS.md
Comment (dblock) [inline MAINTAINERS.md:30]: So we're merging maintainers? :) This can be cleaned up later.

File: abstractions/README.md
Comment (dblock) [inline abstractions/README.md:53]: Delete all the non-abstraction parts. Things like COC are all at the root, we can't have duplicates in various parts of the repo.

**Review by dblock — APPROVED** [review]
Can we please cleanup the README in abstractions/ before merging?

File: abstractions/README.md
Comment (Yury-Fridlyand) [reply abstractions/README.md]: Updated, thanks!

File: MAINTAINERS.md
Comment (Yury-Fridlyand) [reply MAINTAINERS.md:30]: We are merging everything!

**Review by Xtansia — APPROVED** [review]

**Review by dblock — APPROVED** [review]

**Review by VachaShah — APPROVED** [review]

**Pull request merged**
